caas/kubernetes/provider/k8s.go

```diff
@@ -212,13 +212,48 @@ func newK8sBroker(
 			Add(utils.AnnotationModelUUIDKey(isLegacy), modelUUID),
 		isLegacyLabels: isLegacy,
 	}
-	if controllerUUID != "" {
-		// controllerUUID could be empty in add-k8s without -c because there might be no controller yet.
-		client.annotations.Add(utils.AnnotationControllerUUIDKey(isLegacy), controllerUUID)
+	if err := client.ensureNamespaceAnnotationForControllerUUID(controllerUUID, isLegacy); err != nil {
+		return nil, errors.Trace(err)
 	}
 	return client, nil
 }
 
+func (k *kubernetesClient) ensureNamespaceAnnotationForControllerUUID(controllerUUID string, isLegacy bool) error {
+	if len(controllerUUID) == 0 {
+		// controllerUUID could be empty in add-k8s without -c because there might be no controller yet.
+		return nil
+	}
+	ns, err := k.getNamespaceByName(k.namespace)
+	if err != nil && !errors.IsNotFound(err) {
+		return errors.Trace(err)
+	}
+	if ns != nil && !k8sannotations.New(ns.GetAnnotations()).HasAll(k.annotations) {
+		// This should never happen unless we changed annotations for a new juju version.
+		// But in this case, we should have already managed to fix it in upgrade steps.
+		return errors.NewNotValid(nil,
+			fmt.Sprintf("annotations %v for namespace %q must include %v", ns.GetAnnotations(), k.namespace, k.annotations),
+		)
+	}
+	annotationControllerUUIDKey := utils.AnnotationControllerUUIDKey(isLegacy)
+	k.annotations.Add(annotationControllerUUIDKey, controllerUUID)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if ns.GetAnnotations()[annotationControllerUUIDKey] == controllerUUID {
```

Comment:
There's a few references to `ns.GetAnnotations()`, maybe store the value in a variable and refer to that instead.